agent/proxycfg/api_gateway.go

```diff
@@ -125,10 +124,12 @@ func (h *handlerAPIGateway) handleUpdate(ctx context.Context, u UpdateEvent, sna
 			return err
 		}
 	default:
-		return (*handlerUpstreams)(h).handleUpdateUpstreams(ctx, u, snap)
+		if err := (*handlerUpstreams)(h).handleUpdateUpstreams(ctx, u, snap); err != nil {
+			return err
+		}
 	}
 
-	return nil
+	return h.recompileDiscoveryChains(snap)
```

Comment:
Previously, we would recompile the discovery chain every time xDS resources were generated. Instead, this will recompile the discovery chains every time a related resource changes